src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/DecryptorPalOpenSsl.DecodeRecipients.cs

```diff
@@ -0,0 +1,310 @@
+
+            DerSequenceReader originator = encodedKeyAgreement.ReadSequence();
+            byte tag = originator.PeekTag();
+
+            SubjectIdentifierOrKey originatorIdentifierOrKey;
+
+            switch (tag)
+            {
+                case ConstructedSequence:
+                    X509IssuerSerial issuerSerial = ReadIssuerSerial(originator);
+
+                    originatorIdentifierOrKey =
+                        new SubjectIdentifierOrKey(SubjectIdentifierOrKeyType.IssuerAndSerialNumber, issuerSerial);
+
+                    break;
+
+                case DerSequenceReader.ContextSpecificTagFlag:
+                    string skid = ReadSubjectKeyId(originator);
+
+                    originatorIdentifierOrKey =
+                        new SubjectIdentifierOrKey(SubjectIdentifierOrKeyType.SubjectKeyIdentifier, skid);
+
+                    break;
+
+                case ContextSpecific1:
+                    PublicKeyInfo publicKeyInfo = ReadOriginatorPublicKey(originator);
+
+                    originatorIdentifierOrKey =
+                        new SubjectIdentifierOrKey(SubjectIdentifierOrKeyType.PublicKeyInfo, publicKeyInfo);
+
+                    break;
+
+                default:
+                    throw new CryptographicException(SR.Format(SR.Cryptography_Cms_Invalid_Originator_Identifier_Choice, tag));
+            }
+
+            if (encodedKeyAgreement.PeekTag() == ContextSpecific1)
+            {
+                // Skip extra keying material; this one isn't exposed in the API and Framework does
+                // nothing with it.
+                encodedKeyAgreement.SkipValue();
+            }
+
+            AlgorithmIdentifier keyEncryptionAlgo = encodedKeyAgreement.ReadAlgoIdentifier();
+
+            // As there might be more than one element in RecipientEncryptedKeys, we have to create one KeyAgreeRecipientInfo for each
+            // one of the elements. 
+            List<SubjectIdentifier> recipientIdSet = new List<SubjectIdentifier>();
+            List<CryptographicAttributeObject> otherKeyMaterialSet = new List<CryptographicAttributeObject>();
+            List<DateTime> dateSet = new List<DateTime>();
+            List<byte[]> encryptedKeySet = new List<byte[]>();
+
+            ReadRecipientEncryptedKeys(encodedKeyAgreement, recipientIdSet, otherKeyMaterialSet, dateSet, encryptedKeySet);
+            for (int i = 0; i < recipientIdSet.Count; i++)
+            {
+                KeyAgreeRecipientInfoPalOpenSsl kari =
+                    new KeyAgreeRecipientInfoPalOpenSsl(
+                        version,
+                        recipientIdSet[i],
+                        originatorIdentifierOrKey,
+                        otherKeyMaterialSet[i],
+                        keyEncryptionAlgo,
+                        encryptedKeySet[i],
+                        dateSet[i]);
+                recipientInfoStack.Add(new KeyAgreeRecipientInfo(kari));
```

Comment:
While you've gotten rid of one intermediate list here; I was envisioning iterating this once more, and getting rid of the 4 lists declared above.  Either passing in the 3 fixed items, or inlining the ReadRecipientEncryptedKeys to up here.